### DISCUSSION TITLE
Integridade conceitual vai até onde?

### DISCUSSION THREAD

**Discussion by evertonthepaula** [frontendbr/forum#1841] [Perguntas e Respostas]
Estou um com uma dúvida sobre integridade conceitual no front-end, eu compreendo que os módulos do sistema tem que trabalhar em harmonia e de forma coesa, ou seja como se arquiteta o sistema, porém integridade conceitual também está ligada com a forma como o engenheiro escreve o código? 

Por exemplo em React pode-se escrever componentes com Classes ou Funções, integridade conceitual chega tão perto assim da implementação a ponto de definir estas questões? Me parece que sim, isso se for para gerar uma harmonia entre os módulos do sistema, claro que isso é até onde eu entendi integridade conceitual.

"architecture: The fundamental organization of a system embodied in its components, their relationships to each other, and to the environment, and the principles guiding its design and evolution." Fonte: http://www.iso-architecture.org/ieee-1471/defining-architecture.html

Por outro lado o arquiteto que está ligado a manter a integridade não deve se envolver na implementação.

Não estou encontrando nada escrito em pedra sobre este assunto.

**Comment by ninetails** [comment]
> Não estou encontrando nada escrito em pedra sobre este assunto.

Desconhecia este termo, [tanto que tentei assistir um vídeo pra entender](https://youtu.be/o9IiJ7956QM). Pelo resumo do livro pro vestibular, o que entendi tem a ver sim com arquitetura, mas pelo que o cara do vídeo expôs, parece ser algo relacionado a uma cultura muito mais subjetivo e (como diz a palavra) conceitual.

Levando pro lado que entendi, que é muito mais visual, creio que não é exatamente mas é algo muito mais ligado a [Design System](https://github.com/frontendbr/forum/issues?q=is%3Aissue+design+system). Leva em conta ter padrões, _guidelines_ para a criação de _interfaces_.

Um exemplo prático que vi (não vivenciei mas acompanhei meio de perto) é o [Portinari UI](https://po-ui.io/) da TOTVS. Foi criado este repositório de componentes para "unificar" o visual de todos os sitemas web da TOTVS e de terceiros que acoplarem soluções para que o usuário não sinta visualmente que o sistema tem uma cara, um widget tem outra, um módulo tem outro. Isso vc encontra também fortemente nos produtos da Apple e de terceiros integrados.... Google tem o Material UI... por aí vai.

---

Mas comentando sobre implementação não visual, é bom sempre lembrar que [arquitetura de software teve suas bases em idéias de um arquiteto urbano](https://en.wikipedia.org/wiki/Software_design_pattern#History):

> Patterns originated as an architectural concept by Christopher Alexander as early as 1966

Se for levar em questão da programação...

> (...) integridade conceitual também está ligada com a forma como o engenheiro escreve o código?

Sim, mas você vai encontrar menos coisa ainda escritas em pedra. Mesmo porque escrever _Software_ é e deve ser algo livre (não digo _open source_, mas sim inclusiva, qualquer um pode escrever um software desde quem não manja programação e só copia e cola quanto quem se preocupa com qualidade de código).

Sugiro buscar sobre: **linters**, **design patterns**, **Refactoring**, **SOLID**, **Clean Code** e **Clean Architecture** e livros de autores como [Robert C. Martin "Uncle Bob"](https://en.wikipedia.org/wiki/Robert_C._Martin), [Kent Beck](https://en.wikipedia.org/wiki/Kent_Beck) e [Martin Fowler](https://martinfowler.com/aboutMe.html).

---

Ou seja, é possível aplicar idéias de _integridade conceitual_ na implementação e isso traz melhorias na qualidade de código, mas tal como existem _guidelines_ para design como o _Design System_, como o assunto em programação também é subjetivo, tem que ser algo discutido e seguido entre os membros da equipe de desenvolvimento.

Para **frontend**, não há padrões obrigatórios para o desenvolvimento, nem oficialmente recomendável (nem para escrita de código como por exemplo o Golang possui o **gofmt**). Existem diversos padrões de comunidade para escrita de código como [Standard](https://standardjs.com/), [airbnb](https://github.com/airbnb/javascript), mas levando ao nível de arquitetura, saindo de consequência do uso de dado framework que te prende a arquitetura dele, no geral acaba ficando como responsabilidade dos próprios desenvolvedores.
